pdns/syncres.cc

```diff
@@ -2768,6 +2770,19 @@ vState SyncRes::validateRecordsWithSigs(unsigned int depth, const DNSName& qname
           /* that actually does happen when a server returns NS records in authority
              along with the DNSKEY, leading us to trying to validate the RRSIGs for
              the NS with the DNSKEY that we are about to process. */
+          if (name == signer && (type == QType::NSEC || type == QType::NSEC3)) {
+            /* if we are trying to validate the DNSKEY (should not happen here),
+               or more likely NSEC(3)s proving that it does not exist, we have a problem.
+               In that case let's see if the DS does exist, and if it does let's go Bogus
+            */
+            dsmap_t results;
+            vState dsState = getDSRecords(signer, results, false, depth, true);
+            if (dsState == vState::Insecure) {
+              return dsState;
+            }
+            return vState::BogusUnableToGetDNSKEYs;
```

Comment:
What's a bit confusing for me is that we call `getDSRecords` but the error says `BogusUnableToGetDNSKEYs`. The casual reader would expect `BogusUnableToGetDSs`.

Reply:
What's happening is that we are trying to validate a record signed by the `DNSKEY` for `X` while trying to retrieve that exact DNSKEY, which is of course a problem. For some records we can afford to go `Indeterminate`, but for `NSEC` or `NSEC3` records trying to prove that the `DNSKEY` does not exist we can't do that. So we try to see if the zone is indeed signed, which we don't know for sure even though we have some `RRSIG`s because we don't compute the zone cuts beforehand anymore. If we get an `Insecure` result that's fine, there isn't a secure chain from the root to that zone and no `TA` either. But otherwise, we are in a signed Zone, trying to get the corresponding `DNSKEY` and we need that `DNSKEY` to validate records so the problem is not the `DS`.